# AEnet
Alternative splicing (AS), as a crutical promoter of proteomic diversity, constitutes a core source of cellular heterogeneity together with gene expression levels. AS is closely linked to various physiological and pathological states including aging, embryonic development, and tumor progression. Recently, rapid developed single-cell transcriptomics technology has greatly advanced the process of deeply analyzing cellular heterogeneity from gene expression level. However, due to technical limitations such as sequencing depth, dropout events, and batch effects, we have not yet found an ideal approach to comprehensively analyze cellular heterogeneity and its regulatory mechanisms at alternative splicing level. Therefore, this study aims to construct a novel method named alternative splicing network(AEN) combining gene expression levels with AS patterns to analyze cellular heterogeneity, identify novel cell types, and infer the regulatory mechanisms of alternative splicing. By applying the AEN method to gastrula data, pan-cancer T-cell data, and lung cancer tumor cell data, we have successfully revealed the hidden alternative splicing heterogeneity in these cells and identified key alternative splicing events and related splicing factors during cell transformation. The application of AEN method provides new insights into the understanding of cellular heterogeneity and its related physiological and pathological processes.

## The steps for AEnet
AENet consists of six steps: ASP Determination, ASP-Expression Network Construction, Multiple-Sample Integration, Key ASP Identification, ASP Cluster Determination, Cell Type Identification Assistance, and Regulatory Mechanism Prediction. Detailed information for each step is provided below.

### ASP Determination
The goal of the ASP Determination step is to identify alternative splicing patterns within the dataset.

Inputs: The input required for this step: A series of junctions in the dataset from DESJ-detection (https://github.com/liushang17/DESJ-detection). 

Step1: ASP Determination. Exon-exon junctions with the same starting point or endpoint were defined as alternative splicing patterns (ASP).

Outputs: The output of ASP determination is one metadata, including ASPs and the junction composition of ASPs. 

Below is an example command: 

asp <- asp(annj)  # annj is a dataframe with the first column as junctions.

### ASP-Expression Network
At the ASP-Expression Network step, AEN would detect the relationship between ASP usage preference and gene expression dynamic and construct ASP-Expression Network for each sample. 

Inputs. The Gene expression matrix with cell as column, gene as row, and expression as value, as well as the cell metadata with the cell resource information. The Cell-junction CPM matrix and the ASPs metadata from the above step are also needed. 

Step1: detect the relationship between ASPs usage preference and gene expression dynamic. Spearman correlation was calculated between the PSI values of ASPs and gene expression when the number of cells expressing non-missing values of PSI exceeded 10.

Step2: integrate the relationship across all genes and ASPs. All ASPs and genes were represented as nodes in the AEN graph (Alternative Splicing Pattern-Expression Network). ASPs were connected to their significantly associated genes (P value < 0.01), with the line width indicating the strength and direction (positive or negative) of the correlation.

Output. The output of ASP-Expression Network is the AEN network in a file with the four columns: ASP, Gene symbol, correlation coefficient, and p value.

Below is an example command:

cor_list <- step1_cor(exp,mat,asp,ann) # Construct the AEN network for one sample 

cor_list <- mutli(exp,mat,asp,ann) # Construct the AEN network for each sample

### Multiple-Samples Integration
At the multiple samples integration, we would integrate the AEN network from all samples to form the final pan-samples AEN network. 

Inputs. Multiple inputs are required for this step: the AEN network from all samples.

Step1: filter the ASP-Gene pairs in the given batch. AEN filters out ASP-gene pairs that are present in only one sample and excludes those with opposing correlations across different samples. For the remaining pairs, the strength is determined by the number of samples in which the ASP-gene pair is present. The direction of A-E links aligns with the direction observed in single samples.

Output. The output of the steps is the high quality ASP-Gene pairs, with three columns: ASP, Gene Symbol, number of samples, and direction.

Below is an example command:
corm <- merge_cor(cor_list) # Merge the AEN network across multiple samples

### 

## install
devtools::install_github("https://github.com/liushang17/AEN")

## Demo
We showed the application of a demo in the demo.r. The dataset used in the demo.r is shown in the link (https://pan.baidu.com/s/14ZMS2e6fMk3P1CLdOCK6tw . Extraction code ï¼šfyS9).
